### ISSUE TITLE
Bug with Map properties with exposed attributes

### ISSUE THREAD

**Issue by panki** [typestack/class-transformer#319]
```typescript
import 'reflect-metadata';

import { Expose, plainToClass, Type } from 'class-transformer';

export class Skill {
  @Expose() name: string;
}

export class Weapon {
  @Expose() name: string;
  @Expose() range: number;
}

export class Player {
  name: string;

  @Expose()
  @Type(() => Skill)
  skills: Set<Skill>;

  @Expose()
  @Type(() => Weapon)
  weapons: Map<string, Weapon>;
}

const v = {
  name: 'name',
  skills: new Set([
    { name: 'skill1', remove: 1 },
    { name: 'skill2', remove: 2 },
  ]),
  weapons: new Map([
    ['w1', { name: 'weapon1', range: 1, remove: 1 }],
    ['w2', { name: 'weapon2', range: 2, remove: 2 }],
  ]),
};

console.log(
  plainToClass(Player, v, {
    excludeExtraneousValues: true,
  }),
);
```

Output:
```javascript
Player {
  skills: [ Skill { name: 'skill1' }, Skill { name: 'skill2' } ],
  weapons: Map { 'name' => undefined, 'range' => undefined }
}
```

Expected output:
```javascript
Player {
  skills: [ Skill { name: 'skill1' }, Skill { name: 'skill2' } ],
  weapons: Map {
    'w1' => Weapon { name: 'weapon1', range: 1 },
    'w2' => Weapon { name: 'weapon2', range: 2 }
  }
}
```

**Comment by pichillilorenzo** [comment]
You can take a look and try my library [`jackson-js`](https://github.com/pichillilorenzo/jackson-js), which has a lot of decorators and features to easily manipulate JSON, including class transformation as this one. Class transformation is only a little part of what my library can do! As the name implies, it is heavily inspired by the famous Java [FasterXML/jackson library](https://github.com/FasterXML/jackson) and implements almost all of the Java Jackson annotation!

**Comment by randfur** [comment]
We are encountering this issue as well. Using @Expose seems to mess up @Type for Map fields and add the inner type's fields to the map.

**Comment by cvlvxi** [comment]
It appears that when adding @Expose() to the nested class in conjunction with @Type where its tied to a Map<string,T> type does the following

```
    class Weapon {
      @Expose() name: string;
      @Expose() range: number;
    }
      name: string;

      @Expose()
      @Type(() => Skill)
      skills: Set<Skill>;

      @Expose()
      @Type(() => Weapon)
      weapons: Map<string, Weapon>;
    }

    const v = {
      name: "name",
      skills: new Set([
        { name: "skill1", remove: 1 },
        { name: "skill2", remove: 2 },
      ]),
      weapons: new Map([
        ["w1", { name: "weapon1", range: 1, remove: 1 }],
        ["w2", { name: "weapon2", range: 2, remove: 2 }],
      ]),
    };

    console.log(
      plainToClass(Player, v, {
        excludeExtraneousValues: false,
      })
    )
```

Result

```
    Player {
      name: 'name',
      skills: [
        Skill { name: 'skill1', remove: 1 },
        Skill { name: 'skill2', remove: 2 }
      ],
      weapons: Map {
        'w1' => Weapon { name: 'weapon1', range: 1, remove: 1 },
        'w2' => Weapon { name: 'weapon2', range: 2, remove: 2 },
        'name' => undefined,
        'range' => undefined
      }
    }
```

- You can see that it added the Weapon properties as the string keys of the Map which doesn't seem right.. 
- This is with Expose() added and also excludeExtraneousValues set to false
- When we set this to true we get the abnormal behaviour that the above users is experiencing

```
    class Skill {
      @Expose() name: string;
    }

    class Weapon {
      @Expose() name: string;
      @Expose() range: number;
    }

    class Player {
      name: string;

      @Expose()
      @Type(() => Skill)
      skills: Set<Skill>;

      @Expose()
      @Type(() => Weapon)
      weapons: Map<string, Weapon>;
    }

    const v = {
      name: "name",
      skills: new Set([
        { name: "skill1", remove: 1 },
        { name: "skill2", remove: 2 },
      ]),
      weapons: new Map([
        ["w1", { name: "weapon1", range: 1, remove: 1 }],
        ["w2", { name: "weapon2", range: 2, remove: 2 }],
      ]),
    };

    console.log(
      plainToClass(Player, v, {
        excludeExtraneousValues: true,
      })
    );
  });
```

Result

```
    Player {
      skills: [ Skill { name: 'skill1' }, Skill { name: 'skill2' } ],
      weapons: Map { 'name' => undefined, 'range' => undefined }
    }
```

**Comment by NoNameProvided** [comment]
This has been fixed in develop via #405 and be included in the next release.
